source/common/network/filter_manager_impl.cc

```diff
@@ -68,7 +68,8 @@ void FilterManagerImpl::onRead() {
 
 FilterStatus FilterManagerImpl::onWrite() {
   for (const WriteFilterSharedPtr& filter : downstream_filters_) {
-    FilterStatus status = filter->onWrite(buffer_source_.getWriteBuffer());
+    BufferSource::StreamBuffer write_buffer = buffer_source_.getWriteBuffer();
+    FilterStatus status = filter->onWrite(write_buffer.buffer, write_buffer.end_stream);
```

Comment:
Huh.  I was going to ask for a test which verified that the FIN couldn't get ahead of buffered data, but I guess that has to be done on a filter by filter basis.  Hopefully the ASSERT on the write path will catch that for anyone who does it wrong.

It would be nice to have a test filter for integration tests which would (based on filter config)
send data but delay FIN for (short timer), send data + fin immediately, delay data and fin just to make sure the pipeline resumes and there's no problems.  Not sure how hard custom test filters are but I *think* they should be pretty straight-forward.

Reply:
I opted to add some more unit tests on ConnectionImpl for this, because it was much easier to manage with a mock filter.